### PR TITLE
Adding a handle for diagnostics to the IRewriteStep interface

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR adds an additional handle to the IRewriteStep interface that allows compiler plugins to generate diagnostics that will be displayed to the user.

### CODE REVIEW

File: src/QsCompiler/Compiler/CompilationLoader.cs [hunk 1]
```diff
@@ -436,22 +436,44 @@ private Status ExecuteRewriteStep(RewriteSteps.LoadedStep rewriteStep, QsCompila
...
+                    this.LogAndUpdate(ref status, Warning(WarningCode.PreconditionVerificationFailed, new[] { rewriteStep.Name, messageSource }));
+                    return status;
+                }
 
-                var executeTransformation = preconditionPassed && rewriteStep.ImplementsTransformation;
-                var transformationPassed = !executeTransformation || rewriteStep.Transformation(compilation, out transformed);
-                if (!transformationPassed) this.LogAndUpdate(ref status, ErrorCode.RewriteStepExecutionFailed, new[] { rewriteStep.Name, messageSource });
+                var transformationFailed = rewriteStep.ImplementsTransformation && !rewriteStep.Transformation(compilation, out transformed);
+                var postconditionFailed = this.Config.EnableAdditionalChecks && rewriteStep.ImplementsPostconditionVerification && !rewriteStep.PostconditionVerification(transformed);
+                LogDiagnostics();
```
Comment (ScottCarda-MS) [inline src/QsCompiler/Compiler/CompilationLoader.cs]: Do you want to log diagnostics regardless of failure here?

File: src/QsCompiler/Compiler/PluginInterface.cs [hunk 2]
```diff
@@ -1,73 +1,127 @@
...
+        public bool ImplementsPostconditionVerification { get; }
+
         /// <summary>
         /// Verifies whether a given compilation satisfies the precondition for executing this rewrite step. 
         /// <see cref="ImplementsPreconditionVerification"/> indicates whether or not this method is implemented. 
+        /// If the precondition verfication succeeds, then the invocation of an implemented transformation (if any) 
+        /// with the given compilation should complete without throwing an exception. 
+        /// The precondition verification should never throw an exception, 
+        /// but instead indicate if the precondition is satisfied via the returned value. 
+        /// More detailed information can be provided via loggging.
```
Comment (ScottCarda-MS) [inline src/QsCompiler/Compiler/PluginInterface.cs]: spelling : logging

File: src/QsCompiler/Compiler/CompilationLoader.cs (hunk 1 above)
Comment (bettinaheim) [reply src/QsCompiler/Compiler/CompilationLoader.cs]: Yes, I believe we will want to do that.

File: src/QsCompiler/Compiler/PluginInterface.cs (hunk 2 above)
Comment (bettinaheim) [inline src/QsCompiler/Compiler/PluginInterface.cs]:
```suggestion
        /// More detailed information can be provided via logging. 
```

Comment (bettinaheim) [reply src/QsCompiler/Compiler/PluginInterface.cs]: Thanks!

**Pull request merged**